libs/core/executors/include/hpx/executors/fork_join_executor.hpp

```diff
@@ -133,108 +178,101 @@ namespace hpx { namespace execution { namespace experimental {
                 std::size_t const num_threads_;
                 std::size_t const thread_index_;
                 loop_schedule const schedule_;
-                std::atomic<thread_state>& thread_state_;
                 hpx::lcos::local::spinlock& exception_mutex_;
                 std::exception_ptr& exception_;
-                std::chrono::nanoseconds& yield_delay_;
+                std::uint64_t yield_delay_;
 
                 // Changing data for each parallel region.
-                std::atomic<thread_function_helper_type*>&
-                    thread_function_helper_;
-                std::atomic<void*>& element_function_;
-                std::atomic<void const*>& shape_;
-                std::atomic<std::size_t>& size_;
-                std::atomic<void*>& argument_pack_;
+                region_data_type& region_data_;
                 queues_type& queues_;
-                thread_states_type& thread_states_;
 
-                void wait_not_state_this_thread(thread_state state)
+                void set_state_this_thread(thread_state state) noexcept
                 {
-                    hpx::chrono::high_resolution_timer t;
-                    while (
-                        thread_state_.load(std::memory_order_acquire) == state)
-                    {
-                        if (t.elapsed_nanoseconds() > yield_delay_.count())
-                        {
-                            hpx::this_thread::yield();
-                        }
-                    }
+                    region_data_[thread_index_].data_.state_.store(
+                        state, std::memory_order_release);
                 }
 
-                void set_state_this_thread(thread_state state)
+                thread_state get_state_this_thread() const noexcept
                 {
-                    thread_state_.store(state, std::memory_order_release);
+                    return region_data_[thread_index_].data_.state_.load(
+                        std::memory_order_relaxed);
                 }
 
-                void operator()()
+                void operator()() noexcept
                 {
-                    HPX_ASSERT(thread_state_ == thread_state::starting);
+                    HPX_ASSERT(
+                        get_state_this_thread() == thread_state::starting);
                     set_state_this_thread(thread_state::idle);
 
-                    do
-                    {
-                        wait_not_state_this_thread(thread_state::idle);
-                        if (thread_state_ == thread_state::stopping)
-                        {
-                            break;
-                        }
+                    region_data& data = region_data_[thread_index_].data_;
+                    auto state = shared_data::wait_state_this_thread(
+                        data.state_, thread_state::idle, yield_delay_,
+                        std::not_equal_to<>());
```

Comment:
Final comment. Does the region data not have to be loaded _after_ waiting for the state to change and the load should be `acquire`? I would expect the above code to be the "inverse" of https://github.com/STEllAR-GROUP/hpx/blob/29fc9d457587fecd6d40af7f8a3e98ac2fdc0162/libs/core/executors/include/hpx/executors/fork_join_executor.hpp#L515-L523, i.e.
```
... load(state, acquire) ...;
region_data& data = ...;
```
Is there some other memory barrier that ensures that the load can actually be `relaxed`?

Reply:
I see now that `data` is a reference so loading that before the wait should not be an issue. The question about `relaxed` vs `acquire` still stands.

Reply:
Sorry, but which `load()` do you refer to?

Reply:
The one(s) inside `wait_state_this_thread`.

Reply:
Here is my reasoning: the function `wait_state_this_thread` is called in order to wait for a given thread's state to change: either for it to become equal to or to be changed away from the current state. In both cases there are two options: either the condition is already true whenever the `load()` is called (in which case the function returns), or the condition is false (in which case the function will try again). In both scenarios, a `load(relaxed)` is fine - if the condition is true, we know that we can stop waiting and no possible reordering of operations will make it false again. If the condition is false the operation will be retried, even if because of some reordering the actual state was already changed and our `load()` didn't see the state change yet. The next iteration will probably see the change. Do I miss something?

Reply:
Let me add that I probably should flip the use of `equal_to` and `not_equal_to` and change the test conditions inside the function to the opposite. This would better reflect my reasoning laid out above.

Reply:
I actually liked the inverted condition :P but I like your update as well! In any case, my comment had nothing to do with that.

The comment about the `relaxed` load refers to the fact that it doesn't provide any guarantees about reordering other loads around the load of the state. In particular I'm worried about loading the region data here https://github.com/STEllAR-GROUP/hpx/blob/29fc9d457587fecd6d40af7f8a3e98ac2fdc0162/libs/core/executors/include/hpx/executors/fork_join_executor.hpp#L383-L388. An `acquire` load would guarantee that the region data is not reordered to be loaded before reaching the required state. My question is: is there something _else_ that gives the same guarantee if it's using a `relaxed` load?

In practice it's very unlikely that the compiler will rewrite it like that, but I wouldn't want to rely on "very unlikely"...

Reply:
Well, all of those variables are on the same cache line as the state that is loaded atomically. If this isn't guarantee enough, then we will have turn those into atomics as well.

Reply:
Thanks! I may be off here, but I still don't think that's a guarantee. I suppose in the end this function call being behind a function pointer would prevent any chance of the loads being inlined and reordered: https://github.com/STEllAR-GROUP/hpx/blob/29fc9d457587fecd6d40af7f8a3e98ac2fdc0162/libs/core/executors/include/hpx/executors/fork_join_executor.hpp#L214-L217. However, that's again not a guarantee.

Let me ask a different question: did you see a speed-up with `relaxed` vs. `acquire`? I have a vague memory that you mentioned that in some comment or conversation, but can't find it anywhere now.

Reply:
I just wanted to add that it might be easier to figure this out together in Thursday's call...

Reply:
> However, that's again not a guarantee.

Indeed. 

> Let me ask a different question: did you see a speed-up with `relaxed` vs. `acquire`? I have a vague memory that you mentioned that in some comment or conversation, but can't find it anywhere now.

I did not see any speedup (on x86 those are equivalent anyways). I'll change it to `acquire`, then.

Reply:
Thank you. I may be wrong in the end but especially if the performance is the same I'm a lot more comfortable with an `acquire`.

Reply:
Wait, I fear that we've badly misunderstood each other. In https://github.com/STEllAR-GROUP/hpx/pull/5697/commits/020deb82058ad8aa3cde3901fccd409af4082d73 you've changed some `relaxed` memory orderings. I was referring to the loads of the thread state... When setting the state it's set with `release` after writing to other variables. When reading the state it should be read with `acquire` so that reads of the other variables definitely see the writes to them that happened before writing the state.

Reply:
To expand a bit more, this is what I'm going for with this: https://en.cppreference.com/w/cpp/atomic/memory_order#Release-Acquire_ordering. The reads/writes of other variables are ordered appropriately with respect to changes to the thread states (when using `release`/`acquire`).

Reply:
@msimberg I do agree with you now. Thanks for being persistent! This also allows us to make the function pointer in the region data to be non-atomic.

Reply:
Thanks for bearing with me, and more importantly for doing this PR in the first place!